janusgraph-core/src/main/java/org/janusgraph/graphdb/query/index/AbstractIndexSelectionStrategy.java

```diff
@@ -0,0 +1,217 @@
+// Copyright 2020 JanusGraph Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package org.janusgraph.graphdb.query.index;
+
+import javax.annotation.Nullable;
+import com.google.common.base.Preconditions;
+import org.janusgraph.core.JanusGraphElement;
+import org.janusgraph.core.JanusGraphRelation;
+import org.janusgraph.core.PropertyKey;
+import org.janusgraph.core.RelationType;
+import org.janusgraph.core.attribute.Cmp;
+import org.janusgraph.core.schema.JanusGraphSchemaType;
+import org.janusgraph.core.schema.SchemaStatus;
+import org.janusgraph.graphdb.database.IndexSerializer;
+import org.janusgraph.graphdb.internal.OrderList;
+import org.janusgraph.graphdb.query.QueryUtil;
+import org.janusgraph.graphdb.query.condition.*;
+import org.janusgraph.graphdb.query.graph.JointIndexQuery;
+import org.janusgraph.graphdb.types.*;
+import org.janusgraph.graphdb.types.system.ImplicitKey;
+
+import java.util.*;
+import java.util.stream.StreamSupport;
+
+public abstract class AbstractIndexSelectionStrategy {
```

Comment:
This class looks like self sufficient class. Self sufficient classes should not be `abstract`. Now it looks more like Utility class. 
To make it logically abstract I suggest to add `implements IndexSelectionStrategy` but don't implement it in this class directly. Instead, children of this class will need to implement methods from `IndexSelectionStrategy`.
Thus, you don't need to add `implements IndexSelectionStrategy` to any children of this class.